cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java

```diff
@@ -82,69 +110,109 @@
-      CompletableFuture<ProgramController.State> programCompletion = new CompletableFuture<>();
-      ProgramController controller = programRunner.run(program, programOptions);
-      controller.addListener(new AbstractListener() {
-        @Override
-        public void completed() {
-          programCompletion.complete(ProgramController.State.COMPLETED);
-        }
+    Injector injector = Guice.createInjector(createModules(runtimeJobEnv, createCConf(runtimeJobEnv)));
+    LogAppenderInitializer logAppenderInitializer = injector.getInstance(LogAppenderInitializer.class);
+    Deque<Service> coreServices = createCoreServices(injector);
+    startCoreServices(coreServices, logAppenderInitializer);
 
-        @Override
-        public void killed() {
-          programCompletion.complete(ProgramController.State.KILLED);
-        }
+    try {
+      SystemArguments.setLogLevel(programOptions.getUserArguments(), logAppenderInitializer);
+      CConfiguration cConf = injector.getInstance(CConfiguration.class);
+      ProgramRunner programRunner = injector.getInstance(ProgramRunnerFactory.class).create(programId.getType());
 
-        @Override
-        public void error(Throwable cause) {
-          programCompletion.completeExceptionally(cause);
-        }
-      }, Threads.SAME_THREAD_EXECUTOR);
+      // Create and run the program. The program files should be present in current working directory.
+      try (Program program = createProgram(cConf, programRunner, programDescriptor, programOptions)) {
+        CompletableFuture<ProgramController.State> programCompletion = new CompletableFuture<>();
+        ProgramController controller = programRunner.run(program, programOptions);
+        controllerFuture.complete(controller);
+
+        controller.addListener(new AbstractListener() {
+          @Override
+          public void completed() {
+            programCompletion.complete(ProgramController.State.COMPLETED);
+          }
 
-      // Block on the completion
-      programCompletion.get();
+          @Override
+          public void killed() {
+            programCompletion.complete(ProgramController.State.KILLED);
+          }
+
+          @Override
+          public void error(Throwable cause) {
+            programCompletion.completeExceptionally(cause);
+          }
+        }, Threads.SAME_THREAD_EXECUTOR);
+
+        // Block on the completion
+        programCompletion.get();
+      } finally {
+        if (programRunner instanceof Closeable) {
+          Closeables.closeQuietly((Closeable) programRunner);
+        }
+      }
+    } catch (Throwable t) {
+      controllerFuture.completeExceptionally(t);
+      throw t;
     } finally {
-      if (programRunner instanceof Closeable) {
-        ((Closeable) programRunner).close();
+      stopCoreServices(coreServices, logAppenderInitializer);
+    }
+  }
+
+  /**
+   * Stops the running job.
+   */
+  private void stop() {
```

Comment:
could you please add information on javadoc on why this is needed?

Reply:
fixed